app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt

```diff
@@ -0,0 +1,226 @@
+        }
+
+        val handler = Handler(Looper.myLooper()!!)
+        viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
+            override fun onPageSelected(position: Int) {
+                super.onPageSelected(position)
+                handler.removeCallbacks(runnable)
+                handler.postDelayed(runnable, 4000)
+            }
+        })
+        val viewPagerAdapter = ViewPagerAdapter(viewPager, advices, listener)
+        viewPagerAdapter.submitList(advices)
+
+        viewPager.adapter = viewPagerAdapter
+        viewPager.post(runnable)
+        viewPager.offscreenPageLimit = 3
+        viewPager.clipToPadding = false
+        viewPager.clipChildren = false
+        viewPager.getChildAt(0).overScrollMode = RecyclerView.OVER_SCROLL_NEVER
+
+    }
+
+    private fun ItemViewPagerBinding.setUpTransformer(viewPager: ViewPager2) {
+        val transformer = CompositePageTransformer()
+        transformer.addTransformer(MarginPageTransformer(20))
+        transformer.addTransformer { page, position ->
+            val r = 1 - kotlin.math.abs(position)
+            page.scaleY = 0.85f + r * 0.14f
+        }
+        viewPager.setPageTransformer(transformer)
+    }
+
+    abstract class BaseViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
+    class ViewPagerViewHolder(itemView: View) : BaseViewHolder(itemView) {
+        val binding = ItemViewPagerBinding.bind(itemView)
+    }
+
+    class RecentFoodViewHolder(itemView: View) : BaseViewHolder(itemView) {
+        val binding = ItemHorizontalRecipesBinding.bind(itemView)
+    }
+
+    class ChipsViewHolder(itemView: View) : BaseViewHolder(itemView) {
+        val binding = ItemChipsHomeBinding.bind(itemView)
+    }
+
+    class FilteredFoodViewHolder(itemView: View) : BaseViewHolder(itemView) {
+        val binding = ItemVerticalRecipesBinding.bind(itemView)
+    }
+
+
+    override fun getItemCount() = items.size
+
+    override fun getItemViewType(position: Int): Int {
+        return when (items[position].type) {
+            HomeItemType.HEALTHY_VIEW_PAGER -> {
+                HEALTHY_VIEW_PAGER
+            }
+            HomeItemType.HORIZONTAL_RECYCLER -> {
+                RECENT_FOOD
+            }
+            HomeItemType.VERTICAL_RECYCLER -> {
+                HORIZONTAL_RECIPE
+            }
+            HomeItemType.CHIPS_FILTERED -> {
+                CHIPS_FILTER
+            }
+        }
+    }
+
+    fun submitList(newItems: List<HomeItem<Any>>) {
+        val diffResult = DiffUtil.calculateDiff(
+            AppDiffUtil(
+                items,
+                newItems,
+                ::areItemsTheSame
+            )
+        )
+        items = newItems
+        diffResult.dispatchUpdatesTo(this)
+    }
+
+    private fun areItemsTheSame(i: Int, i1: Int, homeItems: List<HomeItem<Any>>): Boolean {
+        return items[i] == homeItems[i1]
+    }
+
+    private class AppDiffUtil<Recipe>(
```

Comment:
If you don't use BaseAdapter then this code can refactor because now it's useless with HOF